Change float to complex for return from ParameterExpression gradient

enavarro51 opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

Fixes #6813

### Details and comments

This fixes an error caused by the ParameterExpression gradient returning a float when there is an imaginary in the expression.

Cryoris commented on qiskit/circuit/quantumcircuit.py:
```diff
@@ -2063,7 +2063,7 @@ def global_phase(self, angle):
         else:
             # Set the phase to the [0, 2π) interval
-            angle = float(angle)
+            angle = complex(angle).real
```
Do we need to support complex numbers here? I'd rather not since they don't make sense for the global phase and instead make sure that we only pass in real numbers further up the stack.

If we have to do this here, could we check if the imaginary part is 0 and only then cast to a real number? Otherwise users will be able to do 
```
qc.global_phase = 1j
```
and it will silently discard the global phase.

enavarro51 replied:
@Cryoris I dug a little deeper here. All the test failures were in the `opflow `tests. They were computing gradients and receiving a complex answer which created `symengine `problems. None of those tests actually used complex gradients, so it seemed simple and prudent to just return a float or complex from `ParameterExpression.gradient` depending on whether there was a non-zero imaginary.

This solves the user entry problem and the current `opflow `tests. The remaining question is, should there be any `opflow `tests that use complex gradients?

Cryoris replied:
@dlasecki will probably add some tests in his work, so far I think we're ok with the current tests.

Cryoris reviewed (commented):
Could you add a test to cover the case of #6813? 🙂

jakelishman reviewed (changes requested):
If the return type of this function changes (I don't feel strongly one way or the other, except that I expect this is not the last time we'll have type-related bugs with `ParameterExpression`), its type hint needs to change too.  Also `complex.imag` is a real number, not a complex - the current `if` check compares it to `0j`, which is a little misleading.

enavarro51 commented:
Thanks, @jakelishman. I changed the `0j `and the type hint and doc. I also added a test for the complex parameter. I agree there may be larger issues here. The tests that failed were calling `opflow.list_ops.list_op.assign_parameters` which assumes floats.

dlasecki commented:
Dear all, I pulled changes from this PR and it seems to be solving the error that my code was producing. Thanks!

Cryoris reviewed (dismissed):
LGTM 👍🏻

woodsp-ibm commented on qiskit/circuit/parameterexpression.py:
```diff
@@ -265,14 +265,15 @@ def _apply_operation(
         return ParameterExpression(parameter_symbols, expr)
 
-    def gradient(self, param) -> Union["ParameterExpression", float]:
+    def gradient(self, param) -> Union["ParameterExpression", complex, float]:
```
Using `complex` alone here as a typehint should suffice. I know things like this have been cleaned up in the past to remove the extra types that are not really needed - its not wrong as such, just not needed. See https://www.python.org/dev/peps/pep-0484/#the-numeric-tower for more info

enavarro51 replied:
Done in https://github.com/Qiskit/qiskit-terra/pull/6847/commits/a7fd512f9f9011482c55269d83dfe93d4cb38bc1.
